external/esp_idf_port/esp32/esp32_queue_api.c

```diff
@@ -0,0 +1,434 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#include <stddef.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+#include <semaphore.h>
+#include <pthread.h>
+#include <errno.h>
+#include <sched.h>
+#include <debug.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <tinyara/arch.h>
+#include <tinyara/clock.h>
+#include "esp_attr.h"
+#include "esp32_queue_api.h"
+#include "mq_tryreceive.h"
+/************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************/
+#define MAX_QUEUE_INFO 10
+
+/************************************************************************
+ * Private Type Declarations
+ ************************************************************************/
+enum {
+	NORMAL = 0,
+	MIDDLE,
+	HIGH,
+} queue_prio_e;
+
+/************************************************************************
+ * Public Variables
+ ************************************************************************/
+const char *mq_name = "mq_wifi";
+static queue_info_t queues_info[MAX_QUEUE_INFO];
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+/****************************************************************************
+ * Name: queue_create_wrapper
+ *
+ * Description:
+ *	This function is the wrapper of mq_open to establish a connection between a named
+ * message queue and the calling task.
+ *
+ * Inputs:
+ *	queue_len - The length of the queue.
+ *	item_size - The size of each node of the queue.
+ *
+ * Return:
+ *	Returns the pionter of the created mqueue, but NULL in a failure.
+ *
+ ****************************************************************************/
+void *IRAM_ATTR queue_create_wrapper(uint32_t queue_len, uint32_t item_size)
+{
+#define NAME_LEN 20
+	char name[NAME_LEN];
+	bool flag = false;
+	uint32_t mq_id = 0;
+
+	for (int i = 0; i < MAX_QUEUE_INFO; i++) {
+		if (!queues_info[i].valid) {
+			flag = true;
+			mq_id = i;
+			break;
+		}
+	}
+	if (!flag) {
+		return NULL;
+	}
+
+	struct mq_attr attr;
+	attr.mq_maxmsg = queue_len;
+	attr.mq_msgsize = item_size;
+	attr.mq_flags = 0;
+	sprintf(name, "%s%d", mq_name, mq_id);
+	/*Invalid param */
+	queues_info[mq_id].mqd_fd_send = mq_open(name, O_RDWR | O_CREAT, 0666, &attr);
+	if (queues_info[mq_id].mqd_fd_send == (mqd_t) ERROR) {
+		return NULL;
+	}
+	queues_info[mq_id].mqd_fd_recv = mq_open(name, O_RDWR | O_CREAT, 0666, &attr);
+	if (queues_info[mq_id].mqd_fd_recv == (mqd_t) ERROR) {
+		return NULL;
```

Comment:
should release mqd_fd_send here before return NULL

Reply:
right, fixed.